### ISSUE TITLE
[ENHANCEMENT] Exporting a volume to image API should respond with the image definition

### ISSUE THREAD

**Issue by drewmullen** [harvester/harvester#5618]
**Is your enhancement related to a problem? Please describe.**

I need to know the name of the image thats created when I export a volume to an image. Currently this API only generates a name using a `image-` prefix and accepts a display name (non unique). It gives me no information (empty `204` reply) on success. 

**There is no way for me to programmatically track the status of an exported image upon creation**

**Describe the solution you'd like**

`/v1/harvester/persistentvolumeclaims/drew/packer-build?action=export` should respond with the image definition, including the name of the new image.

Bonus: it would be nice to specify a `name` or the name-prefix used instead of relying on the default `image-`

**Comment by bk201** [comment]
@WebberHuang1118 please help look into this.

**Comment by WebberHuang1118** [comment]
I had an implementation [that could report the created VM image from volume exporting.](https://github.com/WebberHuang1118/harvester/commit/dc847746bf5c37461fd2ae9f4bb3faace0bc7e7b)
![export_volume](https://github.com/harvester/harvester/assets/19387129/10e05cb2-8bfe-4268-8da9-f5c71841c93a)

however, in the current design, volume actions [do not write any information on the ResponseWriter](https://github.com/harvester/harvester/blob/master/pkg/api/volume/handler.go#L45C1-L56C2).

If we don't want to treat `exporting volume` as a special case and keep the consistency as usual, certain refactor on all volume actions needs to be considered, thanks. @bk201

**Comment by innobead** [comment]
> If we don't want to treat exporting volume as a special case and keep the consistency as usual, certain refactor on all volume actions needs to be considered, thanks. @bk201

If the original response of any volume API is empty, why would adding this cause some refactoring? I thought the response is never handled now since it's not generated.

**Comment by WebberHuang1118** [comment]
Hi @innobead,
In the current design, volume actioners only write `no content status code` in responseWriter, and there is no interface to obtain additional information, [this is implemented in a shared function](https://github.com/harvester/harvester/blob/master/pkg/api/volume/handler.go#L45C1-L56C2) for all actions.

Of course, we can treat exporting volume action as a special case, like [my implementation for a quick verification](https://github.com/WebberHuang1118/harvester/blob/dc847746bf5c37461fd2ae9f4bb3faace0bc7e7b/pkg/api/volume/handler.go#L56-L59). However, IMHO, it breaks the consistency for volume action handlers, and may also affect the source code readability and quality. We should likely have a better interface design for all volume actioners. Thanks.

**Comment by Yu-Jack** [comment]
We don't need to return in the shared function, it depends the first called `.Write` or `.WriteHeader` although it might print some warning message.

```go
func main() {
	m := mux.NewRouter()

	m.Handle("/health", http.HandlerFunc(HealthHandler)).Methods("GET")

	srv := &http.Server{
		Handler: m,
		Addr:    "127.0.0.1:8000"
	}

	log.Fatal(srv.ListenAndServe())
}

// Comment one of them to test it.
func HealthHandler(w http.ResponseWriter, r *http.Request) {
	w.Write([]byte("OK"))
	w.WriteHeader(http.StatusNotImplemented)
	w.WriteHeader(http.StatusNotFound)
}
```

BTW, when I migrate action/link APIs to subresource, I also noticed this topic. We need to union the return format and deal with the status code in one place.

For example:

1. function returns object without error: return 200 and object content
2. function return empty object without error: return 204
3. function return error: depends the error we defined, might be 404, 500 or something.

However, in our current code base, we only return error without object. If we'd like to generate response body with status 200, we need to do like @WebberHuang1118 did.

I think interface could be like this:

```go
func businessLogic(w http.ResponseWriter, r *http.Request) (interface{}, error) {
	return nil, nil // -> 204
	return obj1, nil // -> 200
	return nil, apierror.NewAPIError(validation.NotFound, "Not found resource: {resource name} ") // -> 404
}
```

The `apierror.NewAPIError` is referred from `rancher/apiserver` which we currently used in our action/link APIs.

Then we won't have logic like some functions write its own response logic like this: https://github.com/harvester/harvester/blob/026d49fa55d465c43818068240c2ea8c2a48ea88/pkg/api/vm/handler.go#L458-L478

**Comment by innobead** [comment]
Thanks @Yu-Jack @WebberHuang1118 explanation. The design by Jack sounds good to me.

**Comment by Yu-Jack** [comment]
@bk201 @innobead @WebberHuang1118   I created a PoC draft PR https://github.com/harvester/harvester/pull/6138, you can take a look what my design looks like. Please leave any feedbacks, and I'll keep implementing if we all agree the design, thanks!

**Comment by bk201** [comment]
https://github.com/harvester/harvester/pull/6138 is the first step and looks good. Move to 1.5.0 for more testing time.

**Comment by albinsun** [comment]
Test FAIL due to commit had just merged and not in latest `harvester-v1.4-f5fc452-head`.
Return to ReadyForTesting and will test again later.


## Environment
* `harvester-v1.4-f5fc452-head`
  * Profile: 2 nodes QEMU/KVM (8C/16G/500G per node)
  * ui-source: `Auto`


## Steps 
1. <details>
   <summary> Create a VM
   </summary>

   * VM Running
     ![image](https://github.com/user-attachments/assets/c2a0bcc4-656a-4565-9498-a7e028f08638)
   * VM volume is created
     ![image](https://github.com/user-attachments/assets/54e500c5-814d-46c5-94d8-92c1cc8996f2)
   </details>

1. <details open>
   <summary> Export volume to image by GUI
   </summary>
   
   * Export successfully
     ![image](https://github.com/user-attachments/assets/a63e5d35-8beb-4507-930a-a4effca62dff)
   * Response has image definition :x: 
     > Still reply `204 No Content`
     ![image](https://github.com/user-attachments/assets/e2c6e5c4-fdc3-4a2b-814f-6564ddd9850f)
   </details>

1. <details>
   <summary> Export volume to image by HTTP API
   </summary>

   </details>

**Comment by albinsun** [comment]
## Test Result
Test **PASS**, close as done.


## Environment
* `harvester-v1.4.0-dev-20240918`
  * Profile: 2 nodes QEMU/KVM (8C/16G/500G per node)
  * ui-source: `Auto`

### Prerequisite
1. <details>
   <summary> Create a VM
   </summary>
  
   * VM Running
     ![image](https://github.com/user-attachments/assets/89f4dfd2-a26a-4dc9-a89d-89592e70548c)
   * VM volume is created
     ![image](https://github.com/user-attachments/assets/b349e198-f079-4b55-a6a4-8798c03eaabc)
   </details>

## Steps 
1. <details>
   <summary> Export volume to image by GUI :heavy_check_mark: 
   </summary>
   
   * Export `200 OK`
     ![image](https://github.com/user-attachments/assets/cd3b5be8-4d23-4e33-8e35-3aa94391c111)
   * Response has image definition
     ![image](https://github.com/user-attachments/assets/948a333c-2c77-42f0-a318-4eac14435496)
   </details>

1. <details>
   <summary> Export volume to image by HTTP API :heavy_check_mark: 
   </summary>
   
   * Export `200 OK`
   * Response is VMI
     ![image](https://github.com/user-attachments/assets/259b185c-a444-480e-a7ac-464ebc624f33)
   </details>
   

### QA Check List
- [x] Labels: `not-require/test-plan`, `not-require/release-note`, `require-ui`, `regression`...
- [x] Update corresponding tests ticket lables
